### PR TITLE
Intersect with other shapes

### PR DESCRIPTION
**Pull request by PetervDooren**

### CODE REVIEW

File: src/Shape.cpp [hunk 1]
```diff
@@ -100,11 +100,57 @@ double line_parameter(const Vector3 &D, const Vector3 &v1, const Vector3 &v2, co
...
+    return v1.x*v2.y - v1.y*v2.x;
+}
+
+bool linelineintersect(const Vec2d& v1, const Vec2d& v2, const Vec2d& w1, const Vec2d& w2) {
+    Vec2d ev = v2-v1;
+    Vec2d ew = w2-w1;
+    // check for parallel lines
+    double ev_x_ew = crossProduct(ev, ew);
+    if (ev_x_ew == 0) // lines are parallel
+    {
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]: Be consistent with brackets, following the standard of the rest of the library. So opening bracket on same line.

File: src/Shape.cpp [hunk 2]
```diff
@@ -100,11 +100,57 @@ double line_parameter(const Vector3 &D, const Vector3 &v1, const Vector3 &v2, co
...
+            return false;
+    }
+    else {
+        double tv = crossProduct(w1-v1, ew) / ev_x_ew;
+        double tw = crossProduct(w1-v1, ev) / ev_x_ew;
+        return tv > 0 && tv < 1 && tw > 0 && tw < 1;
+    }
+}
+
+void project3Dto2Dvector(Vec2d& v2d, const Vector3& v3d, int skip){
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]:
```suggestion
void project3Dto2Dvector(Vec2d& v2d, const Vector3& v3d, int skip) {
```

File: src/Shape.cpp [hunk 3]
```diff
@@ -100,11 +100,57 @@ double line_parameter(const Vector3 &D, const Vector3 &v1, const Vector3 &v2, co
     return p_v1 + (p_v2-p_v1) * d_v1/(d_v1-d_v2);
 }
 
+double crossProduct(const Vec2d& v1, const Vec2d& v2) {
+    return v1.x*v2.y - v1.y*v2.x;
+}
+
+bool linelineintersect(const Vec2d& v1, const Vec2d& v2, const Vec2d& w1, const Vec2d& w2) {
```
Comment (MatthijsBurgh) [inline src/Shape.cpp:103]: Would 2 rays not be better? Or create a new line class?

File: src/Shape.cpp [hunk 4]
```diff
@@ -141,19 +188,56 @@ bool Shape::intersect(const Pose3D& pose, const Shape& other) const {
...
+                if (linelineintersect(v1_1_2d, v2_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v1_1_2d, v2_1_2d, v3_2_2d, v1_2_2d)) return true;
+
+                if (linelineintersect(v2_1_2d, v3_1_2d, v1_2_2d, v2_2_2d)) return true;
+                if (linelineintersect(v2_1_2d, v3_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v2_1_2d, v3_1_2d, v3_2_2d, v1_2_2d)) return true;
+
+                if (linelineintersect(v3_1_2d, v1_1_2d, v1_2_2d, v2_2_2d)) return true;
+                if (linelineintersect(v3_1_2d, v1_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v3_1_2d, v1_1_2d, v3_2_2d, v1_2_2d)) return true;
```
Comment (MatthijsBurgh) [inline src/Shape.cpp:217]: Is is very ugly. There should definitely be a better solution for this.

File: src/Shape.cpp [hunk 5]
```diff
@@ -205,7 +289,10 @@ bool Shape::intersect(const Pose3D& pose, const Shape& other) const {
             double t_2l = std::min(t_21,t_22);
 
             if (t_1h > t_2l && t_1l < t_2h)
+            {
+                //std::cout << "Intersection DETECTED"<< std::endl;
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]: Commented code, if you keep the brackets, opening bracket on the same line after a space.

File: src/Shape.cpp [hunk 6]
```diff
@@ -100,11 +100,57 @@ double line_parameter(const Vector3 &D, const Vector3 &v1, const Vector3 &v2, co
     return p_v1 + (p_v2-p_v1) * d_v1/(d_v1-d_v2);
 }
 
+double crossProduct(const Vec2d& v1, const Vec2d& v2) {
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]: Better to add this a member function to `Vec2T` in `math_types.h` I would call it `cross`.

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]
See comments

File: src/Shape.cpp (hunk 4 above)
Comment (PetervDooren) [reply src/Shape.cpp:217]: Made something different for this but I don't know if that makes it better.

File: src/Shape.cpp [hunk 7]
```diff
@@ -94,6 +94,202 @@ bool Shape::intersect(const Vector3& p, const double radius) const {
...
+                    continue;
+
+                // 2D triangle triangle test
+
+                // convert vectors to 2D
+                Vec2d t1_2d[3];
+                Vec2d t2_2d[3];
+
+                double maxN_1 = std::max(std::max(N_1.x, N_1.y), N_1.z);
+                int skip = (maxN_1 == N_1.y) + 2*(maxN_1==N_2.z);
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]:
```suggestion
                int skip = (maxN_1 == N_1.y) + 2*(maxN_1 == N_2.z);
```

File: src/Shape.cpp [hunk 8]
```diff
@@ -94,6 +94,202 @@ bool Shape::intersect(const Vector3& p, const double radius) const {
...
+                project3Dto2Dvector(t1_2d[2], v3_1, skip);
+
+                project3Dto2Dvector(t2_2d[0], v1_2, skip);
+                project3Dto2Dvector(t2_2d[1], v2_2, skip);
+                project3Dto2Dvector(t2_2d[2], v3_2, skip);
+
+                // check all combinations of lines and check for intersections
+                for (int i=0; i!=3; ++i) {
+                    int i_next = i+1;
+                    if (i_next==3) i_next=0;
```
Comment (MatthijsBurgh) [inline src/Shape.cpp]:
```suggestion
                    if (i_next == 3)
                        i_next=0;
```

Comment (MatthijsBurgh) [reply src/Shape.cpp]: I would like to see all these one line ifs to be converted to multi line

**Comment by PetervDooren** [comment]
This PR was not neccessary. While this code does what it is supposed to do it does not solve the problem

**Pull request closed without merging**